builtin/logical/ssh/path_issuers.go

```
@@ -0,0 +1,517 @@
+	}
+
+	return response, nil
+}
+
+func (b *backend) pathWriteIssuerHandler(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	// Since we're planning on updating issuers here, grab the lock so we've
+	// got a consistent view.
+	b.issuersLock.Lock()
+	defer b.issuersLock.Unlock()
+
+	publicKey, privateKey, err := b.keys(d)
+	if err != nil {
+		return handleStorageContextErr(err)
+	}
+
+	// Use the transaction storage if there's one.
+	if txnStorage, ok := req.Storage.(logical.TransactionalStorage); ok {
+		txn, err := txnStorage.BeginTx(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		defer txn.Rollback(ctx)
+		req.Storage = txn
+	}
+
+	sc := b.makeStorageContext(ctx, req.Storage)
+
+	// Create a new issuer entry
+	id, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, fmt.Errorf("error generating issuer's unique identifier: %w", err)
+	}
+	name, err := getIssuerName(sc, d)
+	if err != nil && err != errIssuerNameIsEmpty {
+		return handleStorageContextErr(err)
+	}
+	issuer := &issuerEntry{
+		ID:         id,
+		Name:       name,
+		PublicKey:  publicKey,
+		PrivateKey: privateKey,
+		Version:    1,
+	}
+
+	err = sc.writeIssuer(issuer)
+	if err != nil {
+		return handleStorageContextErr(err, "failed to persist the issuer")
+	}
+
+	response, err := respondReadIssuer(issuer)
+
+	setDefault := d.Get("set_default").(bool)
+	if setDefault {
+		// Update issuers config to set new issuer as the 'default'
+		err = sc.setIssuersConfig(&issuerConfigEntry{DefaultIssuerID: id})
+		if err != nil {
+			// It is not possible to have this error in the transaction, so check
+			// storage type and skip if is a transaction
+			if _, ok := req.Storage.(logical.Transaction); !ok {
+				// Even if the new issuer fails to be set as default, we want to return
+				// the newly submitted issuer with a warning
+				response.AddWarning(fmt.Sprintf("Unable to update the default issuers configuration: %s", err.Error()))
```

I'd probably do this regardless of whether or not we're in a transaction. I'd hate to swallow an error from the transaction layer just because we didn't think it was possible.